### DISCUSSION TITLE
Using the STAGE script to draw an UI?

### DISCUSSION THREAD

**Discussion by melMass** [nxt-dev/nxt_editor#232] [Q&A] [answered]
Hi,

I'm trying to use a simple Qt widget to populate some STAGE or node attributes.
The UI runs perfectly but I can't find a way to stop execution until it's returned/closed.
Does anyone have a suggestion there?
On the other hand I think a basic builtin set of widgets could be very useful ;)

**Comment by ImLucasBrown** [comment, marked as answer]
We don't have documentation for this, and I wouldn't call it feature complete, but we have a builtin toolkit to help with this.
You can right click your layer and select `Reference Builtin Graph > WidgetNodes`. That will bring in some nodes you can instance to make little UIs inside NXT. You'll need to have the `Windows > Workflow Tools` window open to see the widgets. There are some comp issues related to workflow tools, so if something doesn't updated just click the little "comp layer" button in the layer manager.
It is desirable at times to _not_ reset the `STAGE` between workflow tool button clicks, you can control this in via `Execute > Workflow Tools Always Recomp`
Here's a super simple example I just made to show the power of what we have going on: [simple_workflow_tools.zip](https://github.com/nxt-dev/nxt_editor/files/7727951/simple_workflow_tools.zip). We have some users who have taken this to an extreme level, so even with the few comp bugs you might find, it is very capable. 
I wrote very little of the workflow tools implementation, so if you have deeper questions I might be slow in finding the answers.

**Reply by melMass** [reply]
Thanks a lot for the detailed answer and sample! This sound great, I was wondering what the workflow tools where meant for!
